### PR TITLE
Remove redundant sign/zero extension for SIMD broadcasts

### LINKED ISSUE #108820: Unnecessary sign extension in `Vector512.Create(sbyte)` and `Vector512.Create(short)`
### Description

When I was playing around with the `Vector512<sbyte>`, I noticed that `Vector512.Create(sbyte)` first **sign**-extends the value, then broadcasts to the whole vector register.

```csharp
static Vector512<sbyte> Create1(ulong value) => 
    Vector512.Create((sbyte)value);
```
```asm
Program:Create1(ulong):System.Runtime.Intrinsics.Vector512`1[byte] (FullOpts):
       movsx    rsi, sil
       vpbroadcastb zmm0, esi
       vmovups  zmmword ptr [rdi], zmm0
       mov      rax, rdi
       vzeroupper 
       ret
```

Here, sign-extension is unnecessary, as the upper 56 bits of `rsi` would be completely ignored anyway.

A similar thing happens in `Vector512.Create(short)` as well.

```csharp
static Vector512<short> Create2(ulong value) => 
    Vector512.Create((short)value);
```
```asm
Program:Create2(ulong):System.Runtime.Intrinsics.Vector512`1[short] (FullOpts):
       movsx    rsi, si
       vpbroadcastw zmm0, esi
       vmovups  zmmword ptr [rdi], zmm0
       mov      rax, rdi
       vzeroupper 
       ret      
```

As of [uops.info](https://uops.info/table.html?search=mov%20x%20%28R64%2C%20R&cb_lat=on&cb_tp=on&cb_uops=on&cb_ports=on&cb_HSW=on&cb_RKL=on&cb_ADLP=on&cb_ADLE=on&cb_ZEN3=on&cb_ZEN4=on&cb_measurements=on&cb_base=on&cb_bmi=on&cb_mmx=on&cb_x87=on&cb_others=on), in recent Intel CPUs, from Golden Cove and Gracemont microarchitectures onward, the `movzx` instruction would be interpreted as a register rename, so it does not require a clock cycle to execute. The same is true for earlier microarchitectures if the destination register is different from the source register. This is significantly faster than `movsx`, which takes one clock cycle in any case. And `movsx` can be executed in fewer execution ports than `movzx`.

Possible optimization:

- One clock cycle wasted on unnecessary sign extensions can be saved by using `movzx` instead (Intel CPUs only).
  - For example, `movsx rsi, si` → `vpbroadcastw zmm0, esi` can be `movzx rax, si` → `vpbroadcastw zmm0, eax`.
  - Even for AMD CPUs, `movzx` wouldn't take any longer than `movsx` anyway.
- If the value to be broadcast is first processed as data of type 32 bits or greater, eliminating the conversion instruction itself would save instruction decoding resources.
  - For example, `Vector512.Create((sbyte)(rsi >> -8))` would just be `shr rsi, 56` → `vpbroadcastb zmm0, esi`.

A known workaround for this problem is to first use unsigned types to broadcast, then re-interpret the vector register to be one with signed data.
```csharp
static Vector512<sbyte> Create3(ulong value) => 
    Vector512.Create((byte)value).AsSByte();

static Vector512<short> Create4(ulong value) => 
    Vector512.Create((ushort)value).AsInt16();
```
```asm
Program:Create3(ulong):System.Runtime.Intrinsics.Vector512`1[byte] (FullOpts):
       movzx    rsi, sil
       vpbroadcastb zmm0, esi
       vmovups  zmmword ptr [rdi], zmm0
       mov      rax, rdi
       vzeroupper 
       ret      

Program:Create4(ulong):System.Runtime.Intrinsics.Vector512`1[short] (FullOpts):
       movzx    rsi, si
       vpbroadcastw zmm0, esi
       vmovups  zmmword ptr [rdi], zmm0
       mov      rax, rdi
       vzeroupper 
       ret      
```

### Configuration

[Compiler Explorer](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:csharp,selection:(endColumn:40,endLineNumber:16,positionColumn:40,positionLineNumber:16,selectionStartColumn:40,selectionStartLineNumber:16,startColumn:40,startLineNumber:16),source:'using+System%3B%0Ausing+System.Runtime.CompilerServices%3B%0Ausing+System.Runtime.InteropServices%3B%0Ausing+System.Runtime.Intrinsics%3B%0Ausing+System.Numerics%3B%0Ausing+System.Runtime.Intrinsics.X86%3B%0A%0Aclass+Program%0A%7B%0A++++%5BMethodImpl(MethodImplOptions.NoInlining)%5D%0A++++static+Vector512%3Csbyte%3E+Create1(ulong+value)+%3D%3E+%0A++++++++Vector512.Create((sbyte)value)%3B%0A%0A++++%5BMethodImpl(MethodImplOptions.NoInlining)%5D%0A++++static+Vector512%3Cshort%3E+Create2(ulong+value)+%3D%3E+%0A++++++++Vector512.Create((short)value)%3B%0A%0A++++%0A%7D%0A'),l:'5',n:'1',o:'C%23+source+%231',t:'0')),k:46.82726204465335,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:dotnettrunkcsharp,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:2,lang:csharp,libs:!(),options:'--targetos+Linux+--targetarch+x64+--instruction-set+base,sse,sse2,sse3,ssse3,sse4.1,sse4.2,avx,avx2,aes,bmi,bmi2,fma,lzcnt,pclmul,popcnt,avxvnni,evex,avx512f,avx512f_vl,avx512bw,avx512bw_vl,avx512cd,avx512cd_vl,avx512dq,avx512dq_vl,avx512vbmi,avx512vbmi_vl+--Ot+',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+.NET+(main)+(Editor+%231)',t:'0')),k:53.17273795534665,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

### Regression?

Unknown

### Data

[uops.info](https://uops.info/table.html?search=mov%20x%20%28R64%2C%20R&cb_lat=on&cb_tp=on&cb_uops=on&cb_ports=on&cb_HSW=on&cb_RKL=on&cb_ADLP=on&cb_ADLE=on&cb_ZEN3=on&cb_ZEN4=on&cb_measurements=on&cb_base=on&cb_bmi=on&cb_mmx=on&cb_x87=on&cb_others=on)

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/108820

```cs
static Vector512<sbyte> Create1(ulong value) =>
    Vector512.Create((sbyte)value);
```
```diff
; Method Bencha:Create1(ulong)
-      movsx    rdx, dl
       vpbroadcastb zmm0, edx
       vmovups  zmmword ptr [rcx], zmm0
       mov      rax, rcx
       vzeroupper 
       ret      
-; Total bytes of code: 23
+; Total bytes of code: 19
```

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=841197&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by EgorBo** [comment]
@tannergooding @kunalspathak @dotnet/jit-contrib PTAL

**Comment by xtqqczze** [comment]
Diffs show addtional improvement with memory source:

```diff
-       mov      eax, dword ptr [rbp-0x4C]
-       movzx    rax, ax
-       vpbroadcastw ymm0, eax
+       vpbroadcastw ymm0, word  ptr [rbp-0x4C]
```

**Comment by xtqqczze** [comment]
@EgorBo

Missed optimisation with sign extension, e.g.:

```System.PackedSpanHelpers:IndexOf[System.SpanHelpers+DontNegate`1[short],System.PackedSpanHelpers+NopTransform](byref,short,int):int (Instrumented Tier0)```

```asm
        mov      eax, dword ptr [rbp-0x34]
        cwde     
        vpbroadcastb xmm0, eax
```

**Comment by tannergooding** [comment]
I'm confident that changes like this are better:
> Diffs show additional improvement with memory source:
> ```diff
> -       mov      eax, dword ptr [rbp-0x4C]
> -       movzx    rax, ax
> -       vpbroadcastw ymm0, eax
> +       vpbroadcastw ymm0, word  ptr [rbp-0x4C]
> ```


I'm not so confident on changes like:
```csharp
-      movsx    rdx, dl
       vpbroadcastb zmm0, edx
```

`movzx` and `movsx` can avoid stalls otherwise caused by a partial write/merge and so while removing them should be better in the case that a previous instruction initialized either the full 32-bit or 64-bit state, it will likely be worse in cases where something instead only wrote the lower 8/16-bits of state prior to the broadcast.

It might be that the typical JIT patterns are already avoiding such 8/16-bit register writes in most cases, but it would be good to double check this.

File: src/coreclr/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -10356,6 +10356,22 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
 
                                 op1 = node->Op(1);
                                 op1->ClearContained();
+
+                                // Drop GT_CAST if it doesn't change the op1's bits we're about to broadcast
+                                if (op1->OperIs(GT_CAST) && !op1->gtOverflow() && comp->opts.Tier0OptimizationEnabled())
+                                {
+                                    GenTreeCast* cast = op1->AsCast();
+                                    if (!varTypeIsFloating(cast->CastToType()) &&
+                                        !varTypeIsFloating(cast->CastFromType()) &&
```
Comment (tannergooding) [inline src/coreclr/jit/lowerxarch.cpp:10365]: Should this rather check for `varTypeIsInteger` on both?

We have several types (integer, floating, simd, and mask) so a negative test may accidentally include the wrong things

**Review by tannergooding — APPROVED** [review]

**Comment by EgorBo** [comment]
> `movzx` and `movsx` can avoid stalls otherwise caused by a partial write/merge and so while removing them should be better in the case that a previous instruction initialized either the full 32-bit or 64-bit state, it will likely be worse in cases where something instead only wrote the lower 8/16-bits of state prior to the broadcast.
> 
> It might be that the typical JIT patterns are already avoiding such 8/16-bit register writes in most cases, but it would be good to double check this.

I can probably bench it (my bot is currently down), but looks like native compilers don't emit any movs e.g. https://godbolt.org/z/hKrGnz7nj

**Comment by tannergooding** [comment]
> I can probably bench it (my bot is currently down), but looks like native compilers don't emit any movs e.g. https://godbolt.org/z/hKrGnz7nj

Notably this is coming from memory, where its safe to elide.

The consideration would be when it's coming from register and the value was because of a partial register write (i.e. it wrote just `al` or just `ax`). I think this is fine to take overall (hence the sign-off), just something we should watch out for in the perf regressions and potentially consider if tweak is needed as a follow up.

**Comment by xtqqczze** [comment]
@MihuBot

**Pull request merged**